lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -3747,21 +3751,39 @@ class ConvertTritonGPUToLLVM
 
     int numWarps = triton::gpu::TritonGPUDialect::getNumWarps(mod);
 
-    // step 1: Convert FuncOp to LLVMFuncOp via partial conversion
-    // step 2: Allocate for shared memories
-    // step 3: Convert the rest of ops via partial conversion
-    // The reason for a seperation between 1/3 is that, step 2 is out of
+    // step 1: Allocate shared memories and insert barriers
+    // setp 2: Convert SCF to CFG
+    // step 3: Convert FuncOp to LLVMFuncOp via partial conversion
+    // step 4: Convert the rest of ops via partial conversion
+    // The reason for putting step 1 before step 2 is that the membar analysis
+    // currently only supports SCF but not CFG.
+    // The reason for a seperation between 1/4 is that, step 3 is out of
     // the scope of Dialect Conversion, thus we need to make sure the smem
-    // is not revised during the conversion of step 3.
+    // is not revised during the conversion of step 4.
+    Allocation allocation(mod);
+    MembarAnalysis membar(&allocation);
```

Comment:
I didn't see anywhere membar is used, where am i missing?

Reply:
Sorry it's on me.

The analysis function body is applied in `MembarAnalysis`'s constructor, which adds `bar.sync` instructions when necessary. In most of the matmul code, `bar.sync` is added after `wait_group`. That's why it seems a bit confusing. I plan to also rewrite the MembarAnalysis as a pass in the future when making it support CF instead of SCF.

https://github.com/openai/triton/blob/42db3538e4257cac70c6e9c209214bef0a43ca98/include/triton/Analysis/Membar.h#L32

Reply:
Okay I see, MembarAnalysis is not only doing analysis, there should be better names but i think it's ok for this PR.